config_test.go

```diff
@@ -0,0 +1,127 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"runtime"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/packer/packer"
+)
+
+func TestDecodeConfig_basic(t *testing.T) {
+
+	packerConfig := `
+	{
+		"PluginMinPort": 10,
+		"PluginMaxPort": 25,
+		"disable_checkpoint": true,
+		"disable_checkpoint_signature": true
+	}`
+
+	var cfg config
+	err := decodeConfig(strings.NewReader(packerConfig), &cfg)
+	if err != nil {
+		t.Fatalf("error encountered decoding configuration: %v", err)
+	}
+
+	var expectedCfg config
+	json.NewDecoder(strings.NewReader(packerConfig)).Decode(&expectedCfg)
+	if !reflect.DeepEqual(cfg, expectedCfg) {
+		t.Errorf("failed to load custom configuration data; expected %v got %v", expectedCfg, cfg)
+	}
+
+}
+
+func TestLoadExternalComponentsFromConfig(t *testing.T) {
+	packerConfigData, cleanUpFunc, err := generateFakePackerConfigData()
+	if err != nil {
+		t.Fatalf("error encountered while creating fake Packer configuration data %v", err)
+	}
+	defer cleanUpFunc()
+
+	var cfg config
+	cfg.Builders = packer.MapOfBuilder{}
+	cfg.PostProcessors = packer.MapOfPostProcessor{}
+	cfg.Provisioners = packer.MapOfProvisioner{}
+
+	if err := decodeConfig(strings.NewReader(packerConfigData), &cfg); err != nil {
```

Comment:
Thought about consolidating into one wrapper function inside of config `config.LoadData(io.Reader)` or something, but it not sure if it real adds any value. I'm open to suggestions.

Reply:
Yes, that would have been nice to have but I agree it's okay like this 🙂 👍